plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,226 @@
+import PlutusCore hiding (Term)
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin
+import PlutusCore.Data (Data (..))
+import PlutusCore.Generators
 import Test.Tasty
+import Test.Tasty.Hedgehog
+import Type.Reflection
+import Unsafe.Coerce
+
+type Term = PLC.Term TyName Name DefaultUni DefaultFun ()
+
+test_builtinsDon'tThrow :: TestTree
+test_builtinsDon'tThrow =
+    testGroup
+        "Builtins don't throw"
+        $ fmap (\fun -> testProperty (show fun) $ prop_builtinsDon'tThrow fun) List.enumerate
+
+prop_builtinsDon'tThrow :: DefaultFun -> Property
+prop_builtinsDon'tThrow bn = property $ do
+    (args, argStrings) <- unzip <$> forAllNoShow (genArgs bn)
+    mbErr <-
+        liftIO $
+            catch
+                (($> Nothing) . evaluate . runEmitter . runExceptT $ eval args argStrings)
+                (pure . pure)
+    whenJust mbErr $ \(e :: SomeException) -> do
+        let msg =
+                "Builtin function evaluation failed"
+                    <> "Function: "
+                    <> show bn
+                    <> "Arguments: "
+                    <> show argStrings
+                    <> "Error: "
+                    <> show e
+        failWith Nothing msg
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+    eval :: [Term] -> [String] -> MakeKnownM Term
+    eval args0 argStrings = case meaning of
+        BuiltinMeaning _ _ runtime -> go (_broRuntimeScheme runtime) (_broImmediateF runtime) args0
+      where
+        go ::
+            forall n.
+            RuntimeScheme n ->
+            ToRuntimeDenotationType Term n ->
+            [Term] ->
+            MakeKnownM Term
+        go sch f args = case (sch, args) of
+            (RuntimeSchemeArrow sch', a : as) -> do
+                res <- liftEither (f a)
+                go sch' res as
+            (RuntimeSchemeResult, []) -> f
+            (RuntimeSchemeAll sch', _) -> go sch' f args
+            _ -> error $ "Wrong number of args for builtin " <> show bn <> ": " <> show argStrings
+
+-- | Generate arguments to a builtin function based on its `TypeScheme`.
+genArgs :: DefaultFun -> Gen [(Term, String)]
+genArgs bn = sequenceA $ case meaning of
+    BuiltinMeaning tySch _ _ -> go tySch
+      where
+        go :: forall args res. TypeScheme Term args res -> [Gen (Term, String)]
+        go = \case
+            TypeSchemeResult    -> []
+            TypeSchemeArrow sch -> genArg (typeRep @(Head args)) : go sch
+            TypeSchemeAll _ sch -> go sch
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+-- | Generate one argument to a builtin function based on its `TypeRep`.
+genArg :: forall k (a :: k). TypeRep a -> Gen (Term, String)
+genArg tr
+    | Just HRefl <- eqTypeRep tr (typeRep @()) = pure $ mkArg ()
+    | Just HRefl <- eqTypeRep tr (typeRep @Integer) = pure . mkArg =<< genInteger
```

Comment:
I'd write this as `mkArg <$> genInteger`, but whatever.

Reply:
Oh, right. HLint didn't catch this because this hint is disabled in our `.hlint.yaml`..